### ISSUE TITLE
[Slice 2] AI Engine Architecture — Embeddings, Dataset & Model References

### ISSUE THREAD

**Issue by yassinalamelden** [yassinalamelden/ChicFinder#1]
## 🔵 Slice 2 — AI Engine: Real Similarity Search

**Assigned to:** @yassinalamelden  @Amr @MoamenFouad 

---

## 📁 File Ownership

| File | Owner | Status |
|------|-------|--------|
| `ai_engine/embeddings/encoder.py` | Yassin | ✅ Done |
| `ai_engine/embeddings/database_builder.py` | Yassin | ✅ Done |
| `ai_engine/embeddings/vector_store.py` | Yassin | ✅ Done |
| `ai_engine/rag/` | Amr | 🔲 In Progress |
| `ai_engine/llm/` | Moamen | 🔲 In Progress |

---

## 🤖 Model — FashionCLIP

We are using **FashionCLIP** instead of vanilla CLIP for the following reasons:
- Fine-tuned on ~700k Farfetch fashion product images
- Natively understands fashion vocabulary (colors, fabrics, garment types)
- Same 512-dim output — no contract breakage with FAISS index
- No background removal needed — deep pre-trained models handle backgrounds well
- Downloads automatically via HuggingFace, no manual installation needed

**HuggingFace Model:**
🔗 https://huggingface.co/patrickjohncyh/fashion-clip

---

## 🗃️ Dataset — Kaggle Fashion Product Images (Small)

We are using the **Fashion Product Images (Small)** dataset from Kaggle.

**Why this dataset:**
- 44,000 fashion product images covering tops, bottoms, shoes, bags, accessories
- Includes `styles.csv` with metadata (productDisplayName, category, color, gender)
- Clean product shots with white/neutral backgrounds — perfect match for FashionCLIP's training distribution
- `styles.csv` maps directly to Slice 3 metadata fields (Barawy & Gaber)

**Kaggle Dataset:**
🔗 https://www.kaggle.com/datasets/paramaggarwal/fashion-product-images-small

**Kaggle Notebook:** (We are going to use it)
🔗 https://www.kaggle.com/models/jn2clark/fashionclip/PyTorch/fashionsiglip/2

**Reference Notebook (pipeline reference only — not code to copy):**
🔗 https://www.kaggle.com/code/jalesummak/fashion-product-recommendation-system-knn/notebook

---

## ⚙️ Architecture Overview
```
User uploads image
        ↓
ai_engine/embeddings/encoder.py       ← FashionCLIP → 512-d L2-normalized vector
        ↓
ai_engine/embeddings/vector_store.py  ← FAISS IndexFlatIP cosine similarity search
        ↓
Top-K similar fashion items returned to API
```

**Offline (run once before starting API):**
```
data/raw_images/
        ↓
ai_engine/embeddings/database_builder.py  ← builds FAISS index + metadata.json
        ↓
data/embeddings.index + data/metadata.json
```

---

## ✅ Agreed Contracts

- `encode(image_bytes: bytes) -> np.ndarray` — shape `(512,)`, dtype `float32`, L2-norm `1.0`
- `search_similar_items(image_bytes: bytes) -> list[dict]` — drop-in for Slice 1 dummy
- `EMBEDDING_DIM = 512` — fixed, never change without team agreement

---

## ✅ Tasks

- [x] Implement FashionCLIP encoder with singleton pattern
- [x] Implement offline FAISS index builder
- [x] Implement vector store search service
- [x] Test encoder output shape `(512,)` and norm `1.0`
- [ ] Download Fashion Product Images (Small) dataset from Kaggle
- [ ] Run `database_builder.py` to build FAISS index
- [ ] Wire `search_similar_items()` into `api/routers/search.py`
- [ ] Implement `ai_engine/rag/` — Amr
- [ ] Implement `ai_engine/llm/` — Moamen

---

## 📦 Dependencies (Slice 2 relevant)
```
torch
transformers
faiss-cpu
numpy
Pillow
pandas      # for styles.csv parsing in scripts/
```
